Linux packaging docs (#12670)
slint-ui/slint: docs/astro/src/content/docs/guide/platforms/packaging/linux-packaging.mdx

---
title: Linux
description: Package a desktop Slint app for distribution on Linux, using Flatpak
---

{/* cSpell: ignore Flatpaks yourorganization pulseaudio jpegd jpege Metainfo metainfo Flathub launchable */}

import { Aside } from '@astrojs/starlight/components';
import { Tabs, TabItem } from '@astrojs/starlight/components';

> Linux, unlike other platforms, has countless ways to distribute applications. Each of these has its own pros and cons, different supported distros, etc. For this article, we will focus on [Flatpak](https://flatpak.org/). Here is an incomplete list of the alternatives, with the reason that we think they are less well-suited to our purposes here than Flatpak:

Reviewer: Since we're in the guide section, I don't think it's really necessary to justify our choice of Flatpak. A short reference to the alternatives is fine IMO. We're not competing with anyone here :)

- Distro-native package managers (`pacman`, `apt`, `dnf`, etc)
  - Requires building and testing your application on many different systems, in particular different versions of host packages such as glibc
- [Snap](https://snapcraft.io/)
  - Tied to the Canonical ecosystem, usually the best option when distributing a GUI application for Ubuntu and derivatives but far less popular on other distros
- [AppImage](https://appimage.org/)
  - Less well-supported than Snap and Flatpak, the runtime is not shared between applications so the package size is larger compared to Flatpak and Snap. Additionally, AppImage does not provide sandboxing. Since Slint applications can run seamlessly inside Flatpak/Snap's sandbox, so the increased security of a sandbox doesn't require any tradeoff

> Flatpaks are sandboxed, cross-platform, and they are relatively straightforward to create.

Reviewer: This is a positive key phrase :)

<Aside type="note">

> Your application may come with different trade-offs, and so you may want to make another choice.

Reviewer: This could be the box where you could very briefly mention alternatives IMO.

</Aside>

## Prerequisites

> The tools needed are the `flatpak` and `flatpak-builder` tools. Additionally, depending on language, you may need a tool which converts your package manager's lockfile into a set of dependencies that `flatpak-builder` understands. The Flatpak build process is sandboxed, just like Flatpaks themselves, and any tools running inside the build environment do not have network access. Any dependencies that need to be downloaded via the network need to be specified in the Flatpak dependency format.

Reviewer: ```suggestion
The tools needed are the `flatpak` and `flatpak-builder` tools. Additionally, depending on your application's programming language, you may need a tool which converts your build tool's lockfile into a set of dependencies that `flatpak-builder` understands. The Flatpak build process is sandboxed, just like Flatpaks themselves, and any tools running inside the build environment do not have network access. Any dependencies that need to be downloaded via the network need to be specified in the Flatpak dependency format.
```

> Lockfile conversion scripts are available for all languages supported by Slint:

Reviewer: ```suggestion
Lockfile conversion scripts are available for all programming languages supported by Slint:
```

<TabItem label="Rust" icon="seti:rust">

First, you need to run the [`flatpak-cargo-generator.py`](https://github.com/flatpak/flatpak-builder-tools/tree/master/cargo) script, mentioned above. This will create a file named `cargo-sources.json`.

```yaml
modules:
  # The module name is arbitrary, but Flatpak will create a `/run/build/your-module-name` 
  # directory and build the package inside of it, so it should match any `/run/build/..`
  # paths in the build config. It does not need to match the command name.
  - name: slint-test-package
    # This tells Flatpak that you will be writing the build commands out manually, in
    # `build-commands` (see below)
    buildsystem: simple
    build-options:
      # This adds the Rust SDK to your PATH
      append-path: /usr/lib/sdk/rust-stable/bin
      env:
        # Required to keep the Cargo build artifacts inside Flatpak's sandboxed build
        # directory
        CARGO_HOME: /run/build/slint-test-package/cargo
        CARGO_NET_OFFLINE: "true"
        # This is required for Slint's Skia backend, and should match the source below.
        # With the FemtoVG backend, you can remove this line and the relevant source.

>         SKIA_BINARIES_URL: file:///run/build/slint-visual-editor/deps/vendor/skia/skia-binaries-a25a0fdb7d90429aa2d1-x86_64-unknown-linux-gnu-gl-jpegd-jpege-pdf-textlayout-vulkan.tar.gz

Reviewer: This refers to the visual editor. Subsequent for all the skia related bits, I strongly suggest to leave those out of these instructions for now. There's value in supporting Skia here, but I think that could almost go into a separate section.

That said, it's in our interest to support Skia (for our own needs), and for Linux then alternative that I think is easier to support, without the need to hard-code URLs like the above), is to build skia from sources. I think automating that will be easier because instead of the user having to figure out `skia-binaries-a25a0fdb7d90429...` type of URLs, we can use the existing meta-data in [skia-bindings](https://github.com/rust-skia/rust-skia/blob/7307c835f2eb292338fb5e4b45740a07e59dac68/skia-bindings/Cargo.toml#L31) to determine the correct url for the source sources and prepare that.

I tried a local build with skia sources (by setting `SKIA_SOURCE_DIR`) and that works within the sandbox. 

So I see three paths forward:

1. We keep the above as it is. I'd prefer not, because it'll be outdated by the time we release Slint 1.18.
2. We omit skia from the instructions for now and I'll try to create a separate PR to add a section and a helper script for apps that would like to include it and then the build will build it from source.
3. I'll try to finish my experiment with from-source build and we incorporate it in this PR.

Author: My preferred path would be 3, although personally I think this is fine. I give instructions on how to figure out which URL to include.

Not that you're suggesting this, but I definitely don't think that leaving it out is an option since Skia is the best-supported renderer. All of those options are viable though, I wouldn't be unhappy with any of them.

Reviewer: I've included my instructions, so you get to choose :)

    build-commands:
      - cargo --offline fetch --manifest-path Cargo.toml --verbose
      # For the sake of this example, we assume that this build command produces a binary 
      # named `slint-test-package`. If you do not need the Skia renderer, you can use
      # `--features backend-winit,renderer-femtovg` and remove the lines related to Skia 
      # binaries

>       - cargo build --release --offline --features backend-winit,renderer-skia

Reviewer: winit is already in the default set, why specifiy it?

Author: Just for explicitness, I can remove it. I think Skia is the default too though, right?

Reviewer: Skia is not the default renderer in Slint.

Author: Ah, fair enough. On second thoughts, this is just an example project anyway, so those features don't really exist. I could make it `--features slint/renderer-skia`, which will apply to most user projects.

Reviewer: Yes, I think that would be an improvement. But honestly, in a real application you wouldn't toggle features here and there: You research and the decide on a particular feature set and go with it. Either you decide to use Skia, or you don't. And the decision is recorded in your `Cargo.toml`, instead of ending up with a mixture. So I suggest that we default to `cargo build` without dynamically setting features.

Author: Yeah, true. Fair enough, I'll make that change.

      # `renderer-femtovg`, you can remove it
      - type: file
        url: https://github.com/rust-skia/skia-binaries/releases/download/0.99.0/skia-binaries-a25a0fdb7d90429aa2d1-x86_64-unknown-linux-gnu-gl-jpegd-jpege-pdf-textlayout-vulkan.tar.gz
        dest: deps/vendor/skia
        sha256: 097e78d775c9156dc4b070b9cca7008dbab587513ecb1924baf4cf9620f3119b
```

As mentioned in the comments above, to use `renderer-skia` you must include the relevant Skia binaries as an explicit dependency. At the time of writing, this is `0.99.0`. You can use the following command to check the version of `skia-safe` in use (requires `jq` to be installed):

```bash
cargo metadata --format-version 1 | jq -er '.packages | .[] | select(.name == "skia-safe").version'
```

If this is not `0.99.0`, check [the releases page](https://github.com/rust-skia/skia-binaries/tags) for `rust-skia/skia-binaries`, finding the one that has the `x86_64-unknown-linux-gnu` target triple and the `gl-jpegd-jpege-pdf-textlayout-vulkan` feature set. You can also copy the hash and put it in the `sha256` field, mentioned above.

</TabItem>
</Tabs>

## Icons and desktop metadata

> To ensure that your application shows up in the launcher, you need to create a `.desktop` file, along with an icon file. In our case, this will look like so:

Reviewer: ```suggestion
To ensure that your application shows up in the launcher, create a `.desktop` file, along with an icon file. In our case, this will look like so:
```
